taipy/core/data/csv.py

```diff
@@ -219,11 +219,25 @@ def _append(self, data: Any):
                 self._path, mode="a", index=False, encoding=self.properties[self.__ENCODING_KEY], header=False
             )
 
-    def _write(self, data: Any):
-        if isinstance(data, pd.DataFrame):
-            data.to_csv(self._path, index=False, encoding=self.properties[self.__ENCODING_KEY])
+    def __convert_data_to_dataframe(self, data: Any):
+        exposed_type = self.properties[self._EXPOSED_TYPE_PROPERTY]
+        if exposed_type == self._EXPOSED_TYPE_PANDAS and isinstance(data, (pd.DataFrame, pd.Series)):
+            return data
+        elif exposed_type == self._EXPOSED_TYPE_NUMPY and isinstance(data, np.ndarray):
+            return pd.DataFrame(data)
+        elif (
+            isinstance(data, list)
+            and not isinstance(exposed_type, str)
+            and all(isinstance(row, exposed_type) for row in data)
+        ):
```

Comment:
What about splitting this case in two:

- One case where `exposed_type` is a `CustomRow` `dataclass` and `data` is a `dict[CustomRow]`. Then, we apply your code to get column names and convert a data row into a dict of values. In this case, we don't ask the user to implement a Row class inheriting from our base class `BasedCustomExposedType` (The name could be improved, BTW). He just provides a dataclass.

- One case where the user wants to better control the column names. Then we do as we did for mongo data node: We ask the user to provide a CustomRow class that implements encode and decode functions. We can eventually provide a Protocol for type checking.